accepted/2021/net6.0-tfms/net6.0-tfms.md

```diff
@@ -0,0 +1,110 @@
+# .NET 6.0 Target Frameworks
+
+**PM** [Immo Landwerth](https://github.com/terrajobst)
+
+In [.NET 5.0][net5.0] we defined a new syntax for target frameworks (TFM) that
+allows us to avoids tools and humans to use a decoder ring, such as the .NET
+Standard version table, to figure out what is compatible with what. In this
+document, we're listing which TFMs we're adding in .NET 6.0.
+
+## Scenarios and User Experience
+
+TBD
+
+## Requirements
+
+### Goals
+
+* Support the definition of frameworks relevant for Xamarin development
+
+### Non-Goals
+
+* Re-design the TFM syntax
+
+## Stakeholders and Reviewers
+
+* MSBuild / SDK team
+* NuGet team
+* Project system team
+* Xamarin team
+
+## Design
+
+| TFM                | Compatible With                          |
+|--------------------|------------------------------------------|
+| net6.0             | (subsequent version of net5.0)           |
+| net6.0-windows     | (subsequent version of net5.0-windows)   |
+| net6.0-android     | xamarin.android                          |
+|                    | (+everything else inherited from net6.0) |
+| net6.0-ios         | xamarin.ios                              |
+|                    | (+everything else inherited from net6.0) |
+| net6.0-macos       | xamarin.mac                              |
+|                    | (+everything else inherited from net6.0) |
+| net6.0-maccatalyst | xamarin.ios                              |
```

Comment:
Can this compatibility mapping be _removed_ in future major .NET version?  Can `net7.0-maccatalyst` be no longer compatible with `xamarin.ios` while `net6.0-maccatalyst` is?  

I'm not completely sure if this is desirable, just trying to get a sense of if this can go away at some point in the future easily enough.

Can we warn today and err tomorrow?

Reply:
Is the expectation for catalyst and ios to be so highly compatible that the users won't need the special table again to know which ios version is compatible with catalyst version?

Reply:
@jeffhandley will this need a work in platform compact analyzer?

Reply:
> Can this compatibility mapping be _removed_ in future major .NET version? Can `net7.0-maccatalyst` be no longer compatible with `xamarin.ios` while `net6.0-maccatalyst` is?

Keeping the compat relationship rules doesn't cost us, but removing would (because it requires logic to enforce an upper boundary). It would also mess with the developer's intuition on versioning rules. I would say that (1) the old assets are naturally used less because people tend to upgrade and (2) whatever makes us comfortable with people using old assets with issues on .NET 6 should also hold for future versions. I don't think going out of our way to make this no longer work causes more trouble than it is worth.

Reply:
@marek-safar 

> @jeffhandley will this need a work in platform compact analyzer?

Not sure what you're asking. My understanding is that starting with .NET 6, developers either target Catalyst (via `net6.0-maccatalyst`) or iOS (via `net6.0-maccatalyst`) but there is no tooling for building a single binary that can on both (well, ignoring `net6.0`, because that one wouldn't offer any Catalyst/iOS bindings). So in that sense, it doesn't really matter how Catalyst and iOS bindings work. You can have disjoint bindings or a single set of bindings with annotations -- at any point in time, the compiler/analyzer only sees a single set of bindings with version annotations.

And when referencing `xamarin.ios` from Catalyst the platform compat analyzer doesn't do anything because it only looks a the code that is being compiled; it doesn't consider indirect dependencies on platform-specific APIs.

Reply:
It matters for runtime libraries annotations. For example, will `UnsupportedOSPlatform ("ios")` be enough for maccatalyst build? Similarly, are we expecting publicly annotate unsupported API with `UnsupportedOSPlatform ("ios")` to not work also on Catalyst and not to confuse developers?

Reply:
> It matters for runtime libraries annotations. For example, will `UnsupportedOSPlatform ("ios")` be enough for maccatalyst build? Similarly, are we expecting publicly annotate unsupported API with `UnsupportedOSPlatform ("ios")` to not work also on Catalyst and not to confuse developers?

It would become messy if we mixed iOS and Mac Catalyst this way, because there is Mac Catalyst-only API (as there is iOS-only API), and we'd end up with rules like "`UnsupportedOSPlatform ("ios")` applies to Mac Catalyst unless there's a Mac-Catalyst-specific availability attribute, in which case those take precedence."

I think it's best to try to avoid this complication, and just say that availability attributes for iOS do not apply to Mac Catalyst.

Reply:
I believe that's how the tooling works today so we'd need to explicitly block this configuration.

Reply:
I would expect the Xamarin bindings to no use `[UnsupportedOSPlatform]` at all -- I think APIs that only work in iOS would be `[SupportedOSPlatform("ios<version>"]`, APIs that only work in Catalyst be `[SupportedOSPlatform("maccatalyst<version>"]` and APIs that work in both would be marked as `[SupportedOSPlatform("ios<version>"][SupportedOSPlatform("ios<version>"]`.

Reply:
@marek-safar, I've added as section to the spec about platform annotations which should answer your question.

Reply:
@terrajobst I think that should work for Xamarin binding. We'll see if that will be sufficient also for runtime (e.g. different Console support).

Reply:
Good point; cross-platform APIs would follow the Blazor WebAssembly pattern of using `UnsupportedOSPlatform`.